### PR TITLE
[SYCL] Support memory clock rate and memory bus width queries

### PR DESCRIPTION
**Pull request by againull**
E2E test: https://github.com/intel/llvm-test-suite/pull/1386

### CODE REVIEW

**Review by steffenlarsen — CHANGES_REQUESTED** [review]

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -3180,7 +3180,40 @@ pi_result piDeviceGetInfo(pi_device Device, pi_device_info ParamName,
     }
     return ReturnValue(FreeMemory);
   }
+  case PI_EXT_INTEL_DEVICE_INFO_MEMORY_CLOCK_RATE: {
+    // If there are no any memory modules then return 0.
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:3184]: Let's document this behavior in pi.h

File: sycl/doc/extensions/supported/sycl_ext_intel_device_info.md [hunk 2]
```diff
@@ -425,6 +426,80 @@ Then the free device memory  can be obtained using the standard get\_info() inte
       auto FreeMemory = dev.get_info<ext::intel::info::device::free_memory>();
     }
 
+
+# Memory Clock Rate #
+
+A new device descriptor will be added which will provide the maximum clock rate of device's global memory.
```
Comment (smaslov-intel) [inline sycl/doc/extensions/supported/sycl_ext_intel_device_info.md:432]:
```suggestion
A new device descriptor is added which provides the maximum clock rate of device's global memory.
```
?

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 1 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:3184]: And in the .md too, and also explain what is returned when a device has multiple memories

File: sycl/doc/extensions/supported/sycl_ext_intel_device_info.md [hunk 3]
```diff
@@ -18,6 +18,7 @@ The Feature Test Macro SYCL\_EXT\_INTEL\_DEVICE\_INFO will be defined as one of
 | 3     | HW threads per EU device query is supported |
 | 4     | Free device memory query is supported |
 | 5     | Device ID is supported |
+| 6     | Memory clock rate and bus width queries are supported |
```
[the remark this answers is not in the document]
Comment (againull) [reply sycl/doc/extensions/supported/sycl_ext_intel_device_info.md:21]: Thanks for catching this, fixed.

File: sycl/doc/extensions/supported/sycl_ext_intel_device_info.md (hunk 2 above)
Comment (againull) [reply sycl/doc/extensions/supported/sycl_ext_intel_device_info.md]: Fixed.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 1 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Updated pi.h and document.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 4]
```diff
@@ -3180,7 +3180,40 @@ pi_result piDeviceGetInfo(pi_device Device, pi_device_info ParamName,
...
+    if (Device->ZeDeviceMemoryProperties->empty())
+      return ReturnValue(pi_uint32{0});
+    ;
```
[the remark this answers is not in the document]
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Removed, it was added accidentally.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 5]
```diff
@@ -3180,7 +3180,40 @@ pi_result piDeviceGetInfo(pi_device Device, pi_device_info ParamName,
...
+                   const ze_device_memory_properties_t &B) -> bool {
+      return A.maxClockRate < B.maxClockRate;
+    };
+    auto MinIt =
+        std::min_element(Device->ZeDeviceMemoryProperties->begin(),
+                         Device->ZeDeviceMemoryProperties->end(), Comp);
+    return ReturnValue(pi_uint32{MinIt->maxClockRate});
+  }
+  case PI_EXT_INTEL_DEVICE_INFO_MEMORY_BUS_WIDTH: {
+    // If there are no any memory modules then return 0.
```
[the remark this answers is not in the document]
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Fixed.

File: sycl/unittests/kernel-and-program/DeviceInfo.cpp [hunk 6]
```diff
@@ -121,6 +127,52 @@ TEST_F(DeviceInfoTest, GetDeviceFreeMemory) {
...
+  context Ctx{Plt.get_devices()[0]};
+  TestContext.reset(new TestCtx(Ctx));
+
+  device Dev = Ctx.get_devices()[0];
+
+  if (!Dev.has(aspect::ext_intel_memory_clock_rate)) {
+    std::clog << "This test is only for the devices with "
+                 "ext_intel_memory_clock_rate extension support.\n";
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (againull) [reply sycl/unittests/kernel-and-program/DeviceInfo.cpp]: Thanks a lot for these suggestions! Fixed.

**Review by smaslov-intel — APPROVED** [review]

**Review by steffenlarsen — APPROVED** [review]
LGTM!

**Pull request merged**